protocols/medco/private_aggregate_protocol.go

```diff
@@ -0,0 +1,145 @@
+package medco
+
+import (
+	"errors"
+	"github.com/dedis/cothority/log"
+	"github.com/dedis/cothority/network"
+	"github.com/dedis/cothority/sda"
+	"github.com/dedis/cothority/services/medco/libmedco"
+)
+
+// PrivateAggregateProtocolName is the registered name for the private aggregate protocol.
+const PrivateAggregateProtocolName = "PrivateAggregate"
+
+// DataReferenceMessage message sent to trigger an aggregation protocol.
+type DataReferenceMessage struct {
+}
+
+// ChildAggregatedDataMessage contains one node's aggregated data.
+type ChildAggregatedDataMessage struct {
+	ChildData   map[libmedco.GroupingKey]libmedco.CipherVector
+	ChildGroups map[libmedco.GroupingKey]libmedco.GroupingAttributes
+}
+
+type dataReferenceStruct struct {
+	*sda.TreeNode
+	DataReferenceMessage
+}
+
+type childAggregatedDataStruct struct {
+	*sda.TreeNode
+	ChildAggregatedDataMessage
+}
+
+// CothorityAggregatedData is the collective aggregation result.
+type CothorityAggregatedData struct {
+	Groups      map[libmedco.GroupingKey]libmedco.GroupingAttributes
+	GroupedData map[libmedco.GroupingKey]libmedco.CipherVector
+}
+
+func init() {
+	network.RegisterMessageType(DataReferenceMessage{})
+	network.RegisterMessageType(ChildAggregatedDataMessage{})
+	sda.ProtocolRegisterName(PrivateAggregateProtocolName, NewPrivateAggregate)
+}
+
+// PrivateAggregateProtocol performs an aggregation of the data held by every node in the cothority.
+type PrivateAggregateProtocol struct {
+	*sda.TreeNodeInstance
+
+	// Protocol feedback channel
+	FeedbackChannel chan CothorityAggregatedData
+
+	// Protocol communication channels
+	DataReferenceChannel chan dataReferenceStruct
+	ChildDataChannel     chan []childAggregatedDataStruct
+
+	// Protocol state data
+	GroupedData *map[libmedco.GroupingKey]libmedco.CipherVector
+	Groups      *map[libmedco.GroupingKey]libmedco.GroupingAttributes
+}
+
+// NewPrivateAggregate initializes the protocol instance.
+func NewPrivateAggregate(n *sda.TreeNodeInstance) (sda.ProtocolInstance, error) {
+	privateAggregateProtocol := &PrivateAggregateProtocol{
```

Comment:
General remark (don't need to rewrite): for in-method variables it is OK to abbreviate them. So this could be `pap`. For structures we tend to use the full name.